### DISCUSSION TITLE
Question about Algeb diag_eps

### DISCUSSION THREAD

**Discussion by jinningwang** [CURENT/andes#185] [Internal Design] [answered]
Hi Hantao,

I'm curious about the ``diag_eps=True`` in the ``Algeb`` definition, may I ask what's it?

Regards,
Jinning

**Comment by jinningwang** [comment, marked as answer]
Answer from Hantao:

`diag_eps` adds an additional small element to the diagonal element in the Jacobian matrix corresponding to this equation and variable. It is needed when the derivative of this equation w.r.t this variable *may* not be zero to avoid singularity.

For example:
```python
self.vi = Algeb(info='Total input voltages',
                tex_name='V_i',
                unit='p.u.',
                e_str='ue * (-v + vref - WF_y - vi)',
                v_str='-v + vref',
                diag_eps=True,
                )
```
Here if we compute the derivative of ``vi`` (consider ``e_str`` as a function ``g(vi) = 0`` and compute ``dg / d (vi)``), we get ``-ue``. If the device is tripped during the simulation, the connection status flag ``ue`` switches to `0` and thus ``dg / d(vi)=0``. Then we will have 0 appear on the diagonal of the Jacobian matrix, which will cause the matrix to be singular.

Therefore, adding a small value avoids the singularity in such cases.

If `diag_eps = True`, the default `diag_eps` value in the config file will be used. If supplied with a specific value, say `diag_eps = 1e-6`, the specific value will be used. Still, there is no difference in simulation. 

Note:
``ue`` is the connectivity status.

Jinning

**Reply by cuihantao** [reply]
Thanks! Just edited.
